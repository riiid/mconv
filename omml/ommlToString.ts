import { tag } from "../misc/xml.ts";
import {
  OmmlAcc,
  OmmlBar,
  OmmlBorderbox,
  OmmlBox,
  OmmlD,
  OmmlElement,
  OmmlEqarr,
  OmmlF,
  OmmlFunc,
  OmmlGroupchr,
  OmmlLimlow,
  OmmlLimupp,
  OmmlM,
  OmmlNary,
  OmmlOmath,
  OmmlPhant,
  OmmlR,
  OmmlRad,
  OmmlRpr,
  OmmlSpre,
  OmmlSsub,
  OmmlSsubsup,
  OmmlSsup,
  OmmlT,
} from "./model.ts";

export default function ommlToString(node: OmmlElement): string {
  return toStringTable[node.tagName](node as any);
}

const todo = () => "";
const toStringTable: {
  [tagName in OmmlElement["tagName"]]: (
    node: Extract<OmmlElement, { tagName: tagName }>,
  ) => string;
} = {
  "m:oMath": todo,
  "m:acc": todo,
  "m:bar": todo,
  "m:box": todo,
  "m:borderBox": todo,
  "m:d": todo,
  "m:eqArr": todo,
  "m:f": todo,
  "m:func": todo,
  "m:groupChr": todo,
  "m:limLow": todo,
  "m:limUpp": todo,
  "m:m": todo,
  "m:nary": todo,
  "m:phant": todo,
  "m:rad": todo,
  "m:sPre": todo,
  "m:sSub": todo,
  "m:sSubSup": todo,
  "m:sSup": todo,
  "m:r": ommlRToString,
  "m:rPr": todo,
  "m:t": ommlTToString,
};

export function ommlRToString({ tagName, children }: OmmlR): string {
  return tag(tagName, {}, () => {
    // TODO: rpr
    return children.map(ommlTToString).join("");
  });
}

export function ommlTToString({ tagName, xmlSpace, text }: OmmlT): string {
  return tag(tagName, { "xml:space": xmlSpace }, () => text);
}
