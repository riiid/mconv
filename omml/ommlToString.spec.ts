import { assertSnapshot } from "https://deno.land/std@0.139.0/testing/snapshot.ts";
import ommlToString from "./ommlToString.ts";

Deno.test("ommlToString", async (t) => {
  await assertSnapshot(
    t,
    ommlToString({
      tagName: "m:oMath",
      children: [
        { tagName: "m:r", children: [{ tagName: "m:t", text: "23x + 9" }] },
        {
          tagName: "m:sSup",
          ssuppr: { tagName: "m:sSupPr" },
          e: {
            children: [{
              tagName: "m:r",
              children: [{ tagName: "m:t", text: "x" }],
            }],
          },
          sup: {
            children: [{
              tagName: "m:r",
              children: [{ tagName: "m:t", text: "9999" }],
            }],
          },
        },
        { tagName: "m:r", children: [{ tagName: "m:t", text: "= 3" }] },
      ],
    }),
  );

  await assertSnapshot(
    t,
    ommlToString({
      children: [
        {
          children: [
            {
              tagName: "m:t",
              text: "23x + 9",
              xmlSpace: undefined,
            },
          ],
          rpr: undefined,
          tagName: "m:r",
        },
        {
          e: {
            argpr: undefined,
            children: [
              {
                children: [
                  {
                    tagName: "m:t",
                    text: "x",
                    xmlSpace: undefined,
                  },
                ],
                rpr: undefined,
                tagName: "m:r",
              },
            ],
          },
          ssuppr: {
            tagName: "m:sSupPr",
          },
          sup: {
            argpr: undefined,
            children: [
              {
                children: [
                  {
                    tagName: "m:t",
                    text: "9999",
                    xmlSpace: undefined,
                  },
                ],
                rpr: undefined,
                tagName: "m:r",
              },
            ],
          },
          tagName: "m:sSup",
        },
        {
          children: [
            {
              tagName: "m:t",
              text: "= 3",
              xmlSpace: undefined,
            },
          ],
          rpr: undefined,
          tagName: "m:r",
        },
      ],
      tagName: "m:oMath",
    }),
  );
});
