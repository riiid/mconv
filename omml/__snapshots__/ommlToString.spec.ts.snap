export const snapshot = {};

snapshot[`ommlToString 1`] = `"<m:oMath><m:r><m:t>23x + 9</m:t></m:r><m:sSup><m:sSupPr/><m:e><m:r><m:t>x</m:t></m:r></m:e><m:sup><m:r><m:t>9999</m:t></m:r></m:sup></m:sSup><m:r><m:t>= 3</m:t></m:r></m:oMath>"`;

snapshot[`ommlToString 2`] = `"<m:oMath><m:r><m:t>23x + 9</m:t></m:r><m:sSup><m:sSupPr/><m:e><m:r><m:t>x</m:t></m:r></m:e><m:sup><m:r><m:t>9999</m:t></m:r></m:sup></m:sSup><m:r><m:t>= 3</m:t></m:r></m:oMath>"`;
